pkg/controller/user-cluster-controller-manager/kubevirt-vmi-eviction/controller.go

```diff
@@ -0,0 +1,158 @@
+	controllerName   = "kkp-kubevirt-vmi-eviction"
+	machineNamespace = "kube-system"
+)
+
+func init() {
+	if err := kubevirtv1.AddToScheme(scheme.Scheme); err != nil {
+		klog.Fatalf("failed to add kubevirtv1 to scheme: %v", err)
+	}
+	if err := clusterv1alpha1.AddToScheme(scheme.Scheme); err != nil {
+		klog.Fatalf("failed to add kubevirtv1 to scheme: %v", err)
+	}
+}
+
+type reconciler struct {
+	log             *zap.SugaredLogger
+	userClient      ctrlruntimeclient.Client
+	infraClient     ctrlruntimeclient.Client
+	clusterIsPaused userclustercontrollermanager.IsPausedChecker
+}
+
+func Add(ctx context.Context, log *zap.SugaredLogger, userMgr, kubevirtInfraMgr manager.Manager, clusterIsPaused userclustercontrollermanager.IsPausedChecker, clusterName string) error {
+	log = log.Named(controllerName)
+
+	r := &reconciler{
+		log:             log,
+		userClient:      userMgr.GetClient(),
+		infraClient:     kubevirtInfraMgr.GetClient(),
+		clusterIsPaused: clusterIsPaused,
+	}
+
+	c, err := controller.New(controllerName, userMgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller %s: %w", controllerName, err)
+	}
+
+	vmiInformer, err := kubevirtInfraMgr.GetCache().GetInformer(ctx, &kubevirtv1.VirtualMachineInstance{})
+	if err != nil {
+		return fmt.Errorf("failed to get informer for VirtualMachineInstance: %w", err)
+	}
+
+	if err = c.Watch(
+		&source.Informer{Informer: vmiInformer},
+		&handler.EnqueueRequestForObject{},
+		EvacuationNodeNamePredicates(),
+	); err != nil {
+		return fmt.Errorf("failed to watch VirtualMachineInstance: %w", err)
+	}
+
+	return nil
+}
+
+// EvacuationNodeNamePredicates returns a predicate func to only process objects with vmi.Status.EvacuationNodeName != "".
+func EvacuationNodeNamePredicates() predicate.Funcs {
+	return kubermaticpred.Factory(func(o ctrlruntimeclient.Object) bool {
+		if vmi, ok := o.(*kubevirtv1.VirtualMachineInstance); ok {
+			return vmi.Status.EvacuationNodeName != ""
+		}
+		return false
+	})
+}
+
+// Reconcile:
+// - watches the VMI status.evacuationNodeName in the KubeVirt infra cluster.
+// - deletes the corresponding Machine in the user cluster.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("kvvmieviction", request)
+
+	paused, err := r.clusterIsPaused(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to check cluster pause status: %w", err)
+	}
+	if paused {
+		log.Infof("Cluster paused. no reconcile")
+		return reconcile.Result{}, nil
+	}
+
+	vmi := &kubevirtv1.VirtualMachineInstance{}
+	if err := r.infraClient.Get(ctx, request.NamespacedName, vmi); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debug("VirtualMachineInstance not found, returning")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed getting VirtualMachineInstance: %w", err)
+	}
+
+	err = r.reconcile(ctx, log, vmi)
```

Comment:
also could you rename this to `deleteMachine`? That would be more explicit.

Reply:
done